### DISCUSSION TITLE
Need help with Pcode BlockMap, BlockIfElse, BlockGraph

### DISCUSSION THREAD

**Discussion by anzosasuke** [NationalSecurityAgency/ghidra#4314] [Q&A]
I am trying to use pcodes to detect the if/else loop in the program. I came across the BlockIfElse in ghidra. It seems it needs blockgraph and blockmap, but i am having trouble with finding the way to make a blockgraph and blockmap. I have made a basicblock of pcode using highfunction in decompilerutils. I am trying to figure out how to proceed further. Could someone help me with this please? That would be great!

snippet of pcodebasicblock:

```

pcodeblock = res.getHighFunction().getBasicBlocks()
```

**Comment by d-millar** [comment]
@anzosasuke  Up until recently, this place I would have recommended you look was the GraphASTAndFlowScript, which built combined data- and control-flow graphs.  This functionality was recently re-factored into the PCodeCfgGraphTask and PCodeDfgGraphTasks in Ghidra/Features/Decompiler/src/main/java/ghidra/app/plugin/core/decomple/actions.

**Comment by ghidracadabra** [comment]
@anzosasuke It's possible that the method `DecompInterface.structureGraph` is what you are looking for.  There's an example of its use in `DecompilerNestedLayout.performInitialGridLayout`, but that example builds the `ingraph` argument from the `Function` object and not the `HighFunction`.  You'd probably want to build it from the basic blocks from the `HighFunction`.

**Reply by anzosasuke** [reply]
Thanks for that @ghidracadabra . To make a blockgraph instead of a ingraph using the basicblocks, I need to make edges to insert in the graph. There are two methods addBlock() and addEdge(). Since there are no getDestination like methods in pcodeblockbasic and also i couldnot get the address of the particular Pcode Instruction. I could get the address of the varnode using getPCAddress() but not relevant i think. Since I am not able to make any edges,  I am unable to make the block graph as you suggested and get the structure graph. Any help on this is greatly appreciated.

**Reply by d-millar** [reply]
OK, I think (and correct me if I’m wrong) you are looking for the logic underlying the PCodeCfg/DfgGraphTasks.  Take a look at PCodeCombinedGraphTask::buildGraph, particularly the last two loops, and see if the iterator logic is what you are looking for.

**Reply by ghidracadabra** [reply]
@anzosasuke So admittedly this API could use a bit of smoothing over.  

I found some old code where I was experimenting with the graph structuring, here's the method I used to create the BlockGraph:

```
private BlockGraph createInGraph(HighFunction highFunction) {
        Map<PcodeBlockBasic, PcodeBlock> highToGraph = new HashMap<>();
        BlockGraph inGraph = new BlockGraph();
        for (PcodeBlockBasic highBlock : highFunction.getBasicBlocks()) {
            PcodeBlock graphBlock = new BlockCopy(highBlock, highBlock.getStart());
            inGraph.addBlock(graphBlock);
            highToGraph.put(highBlock, graphBlock);
        }

       for (PcodeBlockBasic highBlock : highFunction.getBasicBlocks()) {
            for (int i = 0; i < highBlock.getOutSize(); i++) {
                inGraph.addEdge(highToGraph.get(highBlock), highToGraph.get(highBlock.getOut(i)));
            }
        }
        inGraph.setIndices();
        return inGraph;
    }
```

**Reply by anzosasuke** [reply]
Thanks @d-millar. and thanks @ghidracadabra for the above code. Junggraph is needed. I tried making a junggraph using jungdirectedvisualgraph(). But got the error 'can't instantiate abstract class (ghidra.graph.graphs.JungDirectedVisualGraph)'. I was trying to implement DecompilerNestedLayout.performInitialGridLayout. Am i doing something wrong?

**Comment by dragonmacher** [comment]
Regarding your email question about creating a Jung graph, you should be able to create a `JungDirectedGraph` (not the `JungDirectedVisualGraph` you mentioned).

**Reply by anzosasuke** [reply]
Thanks @dragonmacher . I went ahead with ur idea but got the error like this:
`ingraph.addEdge(highBlock1, highBlock1.getOut(i))
TypeError: addEdge(): 2nd arg can't be coerced to java.util.Collection, edu.uci.ics.jung.graph.util.Pair
`

I also tried the Edge(highBlock1, highBlock1.getOut()) ad similarly with the codeblockEdge, I am still getting this error.

**Reply by decaduto** [reply]
hey @anzosasuke, have you found a way for doing this? I need the same for whiledo and dowhile blocks, do you think is possible to detect them and get the address of the block?
